File: Base/QWTabBarController.m
//
//  ___FILENAME___
//  ___PROJECTNAME___
//  微微卿 制作
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  ___COPYRIGHT___
//
//___COPYRIGHT___
//

#import "___FILEBASENAME___.h"
#import "QWNavitionController.h"
@interface ___FILEBASENAME___ ()

@end

@implementation ___FILEBASENAME___


- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    [self addChildViewControllers];
}
- (void)addChildViewControllers
{
#warning  添加 被TabBarCtl 控制的视图控制器 名称
    //视图控制器名称
    NSMutableArray *array = [NSMutableArray arrayWithArray:@[@"QWBaseViewController",@"QWBaseViewController",@"QWBaseViewController",@"QWBaseViewController"]];
#warning Set TabBarItem Nomal Icon Name
    NSArray *imgArray = @[@"TabBar_home_23x23_",@"TabBar_gift_23x23_",@"TabBar_category_23x23_",@"TabBar_me_boy_23x23_"];
#warning Set TabBarItem selected Icon Name
    NSArray *selectImageArray = @[@"TabBar_home_23x23_selected",@"TabBar_gift_23x23_selected",@"TabBar_category_23x23_selected",@"TabBar_me_boy_23x23_selected"];
#warning Set TabBarItem title
    NSArray *titles = @[@"你可以",@"选择警告",@"修改标题",@"和icon"];
#warning Set NavgationItem title
    NSArray *titles1 = @[@"导航栏标题",@"导航栏标题",@"导航栏标题",@"导航栏标题"];
    for(int i =0;i<array.count;i++)
    {
        UIViewController *vc = [[NSClassFromString(array[i]) alloc] init];
        QWNavitionController *nav = [[QWNavitionController alloc] initWithRootViewController:vc];
        vc.title = titles1[i];
        ;
        nav.tabBarItem.title = titles[i];
        nav.tabBarItem.image = IMG(imgArray[i]);
        nav.tabBarItem.selectedImage = [IMG(selectImageArray[i]) imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
        [array replaceObjectAtIndex:i withObject:nav];
    }
    self.viewControllers = array;
    self.tabBar.tintColor = MAIN_COLOR;
    
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
 
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: Swift/Base/QWTabBarController.swift

//  ___FILENAME___
//  ___PROJECTNAME___
//  微微卿 制作
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  ___FULLUSERNAME___

import UIKit

class QWTabBarController: UITabBarController {
    /// 图标数组 接受外界设置
    var imageArray:Array<String>=[]{
        didSet{
            guard imageArray.count>0 else{
                return
            }
            if childViewControllers.count > 0
            {
                //设置图标
                setItemIcon()
            }
            
        }
    }
    /// 选中 图标数组 接受外界设置
    var selectImageArray:Array<String>=[]{
        didSet{
            guard selectImageArray.count>0 else{
                return
            }
            if childViewControllers.count > 0
            {
                //设置选中图标
                setSelectItemIcon()
            }
            
        }
    }
    /// 视图控制器数组 接受外界设置
    var VCArray:Array<AnyObject>?{
        didSet{
            guard VCArray?.count>0 else{
                return
            }
            if(titleArray.count<=0)
            {
                for i in 0..<Int((VCArray?.count)!)
                {
                    titleArray.append("标题\(i)")
                }
            }
            if(imageArray.count<=0)
            {
                imageArray = ["TabBar_home_23x23_","TabBar_gift_23x23_","TabBar_category_23x23_","TabBar_me_boy_23x23_"]
            }
            if(selectImageArray.count<=0)
            {
                selectImageArray = ["TabBar_home_23x23_selected","TabBar_gift_23x23_selected","TabBar_category_23x23_selected","TabBar_me_boy_23x23_selected"]
            }
            //添加子视图控制器
            addChildViewControllers()
        }
    }
    /// 标题数组 接受外界设置
    var titleArray:Array<String> = []{
        didSet{
            guard titleArray.count>0 else{
                return
            }
            if childViewControllers.count > 0
            {
                //设置标题
                setTitles()
            }
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view.backgroundColor = UIColor.whiteColor()
        
    }
    func addChildViewControllers(){
        //防止外界多次设置造成错乱 移除
        for vc in childViewControllers
        {
            vc.removeFromParentViewController()
        }
        //类名
        let array = self.VCArray
        for i in 0..<array!.count
        {
            let cName = array![i]
            //动态获取命名空间
            //(CFBundleExecutable这个键对应的值就是项目名称,也就是命名空间)
            let nameSpace = NSBundle.mainBundle().infoDictionary!["CFBundleExecutable"] as! String
            //将字符串转化为类
            //默认情况下,命名空间就是项目名称,但是命名空间是可以修改的
            let cls:AnyClass? = NSClassFromString(nameSpace + "." + (cName as! String))
            //通过类创建对象
            //将anyClass转换为指定的类型
            let viewControllerCls = cls as! UIViewController.Type
            //通过class创建对象
            let vc = viewControllerCls.init()
            //包装导航栏
            let navCtl = QWNavitionController.init(rootViewController:vc)
            addChildViewController(navCtl)
            //                navCtl.tabBarItem.selectedImage = UIImage(named:imgName+"selected")
            
        }
        //设置选项卡标题
        setTitles()
        //设置图标
        setItemIcon()
        //设置选中图标
        setSelectItemIcon()
        self.tabBar.tintColor = Main_Color
        
    }
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}
//设置 元素
extension QWTabBarController{
    
    
    
    /**
     设置标题
     */
    func setTitles(){
        
        let count = childViewControllers.count>titleArray.count ? titleArray.count:childViewControllers.count
        for index in 0..<count
        {
            let navVC  = childViewControllers[index] as? QWNavitionController
            navVC!.tabBarItem.title = titleArray[index]
            let vc = navVC?.topViewController
            vc?.navigationItem.title = titleArray[index]
        }
        
    }
    /**
     设置图标
     */
    func setItemIcon(){
        let count = childViewControllers.count>titleArray.count ? titleArray.count:childViewControllers.count
        
        for index in 0..<count
        {
            let navVC  = childViewControllers[index] as? QWNavitionController
            navVC!.tabBarItem.image = IMG(imageArray[index])
        }
    }
    func setSelectItemIcon(){
        let count = childViewControllers.count>titleArray.count ? titleArray.count:childViewControllers.count
        for index in 0..<count
        {
            let navVC  = childViewControllers[index] as? QWNavitionController
            navVC!.tabBarItem.selectedImage = IMG(imageArray[index])
        }
        
    }
    
}

